pkg/client/didexchange/client.go

```diff
@@ -132,22 +134,109 @@ func (c *Client) startServiceEventListener() error {
 		return fmt.Errorf("didexchange message event registration failed: %w", err)
 	}
 
-	// TODO https://github.com/hyperledger/aries-framework-go/issues/199 - Generate Client events
-	// for now, auto execute the actions
+	// listen for action event and message events
 	go func() {
-		err := didexchange.AutoExecuteActionEvent(c.actionCh)
-		if err != nil {
-			logger.Errorf("auto action event execution failed: %s", err)
+		for {
+			select {
+			case e := <-c.actionCh:
+				// assigned to var as lint fails with : Using a reference for the variable on range scope (scopelint)
+				msg := e
+				c.handleActionEvent(&msg)
+			case e := <-c.msgCh:
+				// assigned to var as lint fails with : Using a reference for the variable on range scope (scopelint)
+				msg := e
+				c.handleMessageEvent(&msg)
+			}
 		}
 	}()
 
-	go func() {
-		for e := range c.msgCh {
-			// TODO https://github.com/hyperledger/aries-framework-go/issues/199 - Generate Client events
-			// for now, log the messages
-			logger.Infof("message event received : type=%d state=%s", e.Type, e.StateID)
+	return nil
+}
+
+// RegisterActionEvent on DID Exchange protocol messages. The events are triggered for incoming exchangeRequest,
+// exchangeResponse and exchangeAck message types. The consumer need to invoke the callback to resume processing.
+// Only one channel can be registered for the action events. The function will throw error if a channel is already
+// registered. The AutoExecuteActionEvent() function can be used to automatically trigger callback function for the
+// event.
+func (c *Client) RegisterActionEvent(ch chan<- dispatcher.DIDCommAction) error {
+	c.actionEventlock.Lock()
+	defer c.actionEventlock.Unlock()
+
+	if c.actionEvent != nil {
+		return errors.New("channel is already registered for the action event")
+	}
+
+	c.actionEvent = ch
+
+	return nil
+}
+
+// UnregisterActionEvent on DID Exchange protocol messages. Refer RegisterActionEvent().
+func (c *Client) UnregisterActionEvent(ch chan<- dispatcher.DIDCommAction) error {
+	c.actionEventlock.Lock()
+	defer c.actionEventlock.Unlock()
+
+	if c.actionEvent != ch {
+		return errors.New("invalid channel passed to unregister the action event")
+	}
+
+	c.actionEvent = nil
+
+	return nil
+}
+
+// RegisterMsgEvent on DID Exchange protocol messages. The message events are triggered for state transitions. Client
+// will not expect any callback on these events unlike Action events.
+func (c *Client) RegisterMsgEvent(ch chan<- dispatcher.StateMsg) error {
+	c.msgEventsLock.Lock()
+	c.msgEvents = append(c.msgEvents, ch)
```

Comment:
I wonder if multiple channels will become registered to a particular client.
(I understand why protocol services have multiple listeners - it is less clear that Clients do).

Reply:
Since these are just stream listeners and client doesn't expect anything back. So, providing support for multiple listeners won't harm. You have anything specific in mind ?

Reply:
@rolsonquadras Just wondering if there is unnecessary complexity :).

Reply:
@troyronda Created an issue https://github.com/hyperledger/aries-framework-go/issues/395 to discuss this further.